File: cpp/src/card_tracking/detection_tracker.cpp
#include <algorithm>
#include <cstddef>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <optional>
#include <setdetect/card_detection/async_detector.hpp>
#include <setdetect/card_tracking/detection_tracker.hpp>
#include <setdetect/set_game/game.hpp>
#include <vector>

namespace setdetect::card_tracking {

auto propagate_detection(const img_proc::SparsePropagator& propagator, int from_frame_id,
                         const card_detection::Detection& detection) -> card_detection::Detection {
    card_detection::Detection out;
    out.matches_arrangement = detection.matches_arrangement;

    if (detection.detected_cards.empty()) {
        return out;
    }

    std::vector<img_proc::Pt2f> pts;
    pts.reserve(detection.detected_cards.size() * 4);
    for (const card_detection::DetectedCard& dc : detection.detected_cards) {
        for (const card_detection::Point2d& corner : dc.corners) {
            pts.push_back({static_cast<float>(corner.x), static_cast<float>(corner.y)});
        }
    }

    if (!propagator.propagate(from_frame_id, pts)) {
        return out;
    }

    out.detected_cards.reserve(detection.detected_cards.size());
    for (size_t i = 0; i < detection.detected_cards.size(); ++i) {
        card_detection::DetectedCard dc = detection.detected_cards[i];
        for (size_t k = 0; k < 4; ++k) {
            const img_proc::Pt2f& p = pts[(i * 4) + k];
            dc.corners[k] = {static_cast<double>(p.x), static_cast<double>(p.y)};
        }
        if (const std::optional<set_game::Card> card = card_detection::match_card(dc)) {
            out.matches.emplace_back(dc, *card);
        }
        out.detected_cards.push_back(dc);
    }
    return out;
}

auto scale_detection(const card_detection::Detection& detection, double factor) -> card_detection::Detection {
    card_detection::Detection out = detection;
    for (card_detection::DetectedCard& dc : out.detected_cards) {
        for (card_detection::Point2d& corner : dc.corners) {
            corner.x *= factor;
            corner.y *= factor;
        }
    }
    for (auto& [dc, card] : out.matches) {
        for (card_detection::Point2d& corner : dc.corners) {
            corner.x *= factor;
            corner.y *= factor;
        }
    }
    return out;
}

struct DetectionTracker::Impl {
    img_proc::SparsePropagator propagator;
    card_detection::AsyncCardDetector async_detector;
    int max_side;
    std::optional<card_detection::Detection> last_detection;
    int last_detection_frame_id = -1;

    Impl(const card_detection::CardDetector& detector, const img_proc::SparsePropagatorParams& params, int max_side_)
        : propagator(params), async_detector(detector), max_side(max_side_) {}
};

DetectionTracker::DetectionTracker(const card_detection::CardDetector& detector,
                                   const img_proc::SparsePropagatorParams& propagation_params, int max_side)
    : impl_(std::make_unique<Impl>(detector, propagation_params, max_side)) {}

DetectionTracker::~DetectionTracker() = default;

auto DetectionTracker::detect(const cv::Mat& img_rgb) -> std::optional<card_detection::Detection> {
    // Resize down to the working resolution once (only if larger); everything below
    // operates on `working`, so a fresh detection's and a propagated detection's corners
    // always live in the same coordinate space. INTER_AREA matches Python's
    // cli/card_tracking.py resolution-limiting resize.
    cv::Mat resized;
    const cv::Mat* working = &img_rgb;
    double to_orig = 1.0;
    const int longest = std::max(img_rgb.rows, img_rgb.cols);
    if (longest > impl_->max_side) {
        const auto scale = static_cast<float>(impl_->max_side) / static_cast<float>(longest);
        cv::resize(img_rgb, resized, cv::Size(), scale, scale, cv::INTER_AREA);
        working = &resized;
        to_orig = static_cast<double>(img_rgb.rows) / static_cast<double>(resized.rows);
    }

    cv::Mat gray;
    cv::cvtColor(*working, gray, cv::COLOR_RGB2GRAY);
    const img_proc::PropagationStep step = impl_->propagator.next_img(gray);

    impl_->async_detector.submit(step.frame_id, *working);

    if (const std::optional<card_detection::AsyncCardDetector::Result> result = impl_->async_detector.poll()) {
        impl_->last_detection = propagate_detection(impl_->propagator, result->frame_id, result->detection);
        impl_->last_detection_frame_id = step.frame_id;
    } else if (const std::optional<card_detection::Detection>& prev = impl_->last_detection; prev.has_value()) {
        impl_->last_detection = propagate_detection(impl_->propagator, impl_->last_detection_frame_id, *prev);
        impl_->last_detection_frame_id = step.frame_id;
    }

    if (const std::optional<card_detection::Detection>& last = impl_->last_detection; last.has_value()) {
        if (to_orig == 1.0) {
            return last;
        }
        return scale_detection(*last, to_orig);
    }
    return std::nullopt;
}

} // namespace setdetect::card_tracking

File: py/src/setdetect/card_tracking/detection_tracker.py
import typing as tp

import cv2
import numpy as np

from setdetect.card_detection.async_detector import AsyncCardDetector
from setdetect.card_detection.detector import DetectedCard, Detection, _match_card
from setdetect.card_tracking.sparse_propagator import SparsePropagator


class DetectionTracker:
    """Combined async card detection and card corner propagation."""

    def __init__(
        self, detection_kwargs: dict[str, tp.Any] | None = None, propagation_kwargs: dict[str, tp.Any] | None = None
    ) -> None:
        # parameters
        if detection_kwargs is None:
            detection_kwargs = {}
        if propagation_kwargs is None:
            propagation_kwargs = {}

        # state
        self._epoch = 0
        self._detection_frame_id: int = -1
        self._detection: Detection | None = None

        # detector and propagator
        self._async_detector = AsyncCardDetector(detection_kwargs=detection_kwargs)
        self._propagator = SparsePropagator(**propagation_kwargs)

    def open(self) -> None:
        self._async_detector.open()

    def close(self) -> None:
        self._async_detector.close()

    def __enter__(self) -> "DetectionTracker":
        self.open()
        return self

    def __exit__(self, exc_type: object, exc_value: object, traceback: object) -> None:
        self.close()

    def clear(self) -> None:
        self._epoch += 1
        self._detection_frame_id = -1
        self._detection = None
        self._async_detector.drain()
        self._propagator.clear()

    def _propagate_detection(self, detection: Detection, from_frame_id: int) -> Detection:
        cards = []
        for ddc in detection.detected_cards:
            dc_corners = self._propagator.propagate(from_frame_id, ddc.corners)
            if dc_corners is None:
                continue
            dc = DetectedCard(
                corners=dc_corners,
                corner_visibility=ddc.corner_visibility,
                count_probs=ddc.count_probs,
                color_probs=ddc.color_probs,
                shape_probs=ddc.shape_probs,
                fill_probs=ddc.fill_probs,
            )
            cards.append(dc)
        return Detection(
            detected_cards=cards,
            matches=[(dc, card) for dc in cards if (card := _match_card(dc)) is not None],
            matches_arrangement=detection.matches_arrangement,
        )

    def detect(self, img_rgb: np.ndarray) -> Detection | None:
        img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_RGB2GRAY)
        propagation_step = self._propagator.next_img(img_gray)
        self._detection_frame_id = propagation_step.frame_id

        self._async_detector.submit(propagation_step.frame_id, self._epoch, img_rgb)
        delayed_detection = self._async_detector.poll(self._epoch)

        if delayed_detection is not None:
            self._detection = self._propagate_detection(delayed_detection.detection, delayed_detection.frame_id)
        elif self._detection is not None:
            self._detection = self._propagate_detection(self._detection, self._detection_frame_id)
        else:
            self._detection = None
        return self._detection


class TimedDetectionTracker(DetectionTracker):
    def __init__(
        self, detection_kwargs: dict[str, tp.Any] | None = None, propagation_kwargs: dict[str, tp.Any] | None = None
    ) -> None:
        super().__init__(detection_kwargs, propagation_kwargs)
        self.times = {}

    def detect(self, img_rgb: np.ndarray) -> Detection | None:
        import time

        self.times = {}

        t0 = time.perf_counter()
        img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_RGB2GRAY)
        self.times["cvtColor"] = time.perf_counter() - t0

        t0 = time.perf_counter()
        propagation_step = self._propagator.next_img(img_gray)
        self._detection_frame_id = propagation_step.frame_id
        self.times["next_img"] = time.perf_counter() - t0

        t0 = time.perf_counter()
        self._async_detector.submit(propagation_step.frame_id, self._epoch, img_rgb)
        self.times["submit"] = time.perf_counter() - t0

        t0 = time.perf_counter()
        delayed_detection = self._async_detector.poll(self._epoch)
        self.times["poll"] = time.perf_counter() - t0

        t0 = time.perf_counter()
        if delayed_detection is not None:
            self._detection = self._propagate_detection(delayed_detection.detection, delayed_detection.frame_id)
            self.times["propagate detection"] = time.perf_counter() - t0
        elif self._detection is not None:
            self._detection = self._propagate_detection(self._detection, self._detection_frame_id)
            self.times["propagate one frame"] = time.perf_counter() - t0
        else:
            self._detection = None
            self.times["propagate nothing"] = time.perf_counter() - t0
        return self._detection
